### PR TITLE
Implement UI for Add Non-UM Users feature (#246)

### LINKED ISSUE #246: Implement FE for Add Non-Umich User

### PR DESCRIPTION
**Pull request by ssciolla**
This PR introduces the UI portion of a new feature, "Add Non-UM Users", the foundation of which is the `AddNonUMUsers.tsx` page component. The PR aims to resolve #246.

Other new components:
- `BulkApiErrorContent`: Component encapsulating logic regarding display of error content resulting from a bulk process with multiple possible Canvas errors
- `BulkEnrollExternalUserConfirmationTable`: As in other workflow-oriented features, this feature's confirmation table renders the user's CSV content for review.
- `CreateSelectSectionWidget`: Composite component combining `CreateSectionWidget` and `SelectSectionorWidget` with some explanatory text (can be modified later to help with #284)
- `InlineErrorAlert`: Temporary inline alert implementation mimicking [Material UI 5's `Alert`](https://mui.com/components/alert/), which we can adopt once we migrate to Material UI 5 (see #220).
- `MultipleUserEnrollmentWorkflow`: Stateful component encapsulating the bulk CSV workflow portion of the feature.
- `MethodSelect`: Stateless generic component allowing user to select one of multiple feature input options (I will use this again later with #278).
- `RoleSelect`: Stateless component allowing the selection of one of a handful of specified Canvas roles/enrollment types. 
- `UserEnrollmentForm`: Stateful component encapsulating the single external user form and its success state
- `WorkflowStepper`: Thin wrapper around the Material UI `Stepper` to prevent repetition of style and child component  (`Step`, `StepLabel`) references.
- `ValidatedFormField`: Stateless component and wrapper around Material UI `TextField` handling validation logic display and styling.

Other changes:
- I expanded and tweaked existing validation logic, introducing `rowValidation` with a duplicate validation class that can be used by other validators and `enrollmentValidators`, which is now used by processing adding internal and external users.
- I added role fetching and conversion logic to `client/models/canvas.ts`. More can probably be done to simplify later.

To Do List
- [x] Adjust allowed enrollment role logic
- [x] Validate CSV rows for duplicate email
- [x] Improve error handling strategy in `UserEnrollmentForm`
- [x] Check/correct use of `aria-live`

### CODE REVIEW

**Comment by pushyamig** [comment]
is there any reason why you chose to get the section data during the landing page for Add Non-UM user? Probably to support both features load once than 2 time before the user selection. 

I think this is fine, but felt that adding a spinner saying loading section data when there is nothing to show (with sections) for that view. Felt bit odd. But this just observation. No work is needed.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/CreateSelectSectionWidget.tsx:45]: may be a small space between the text and textbox. I know the PR is still a draft. But just shared my observation. 
Screen Shot 2021-12-17 at 4 32 42 PM
?

**Comment by pushyamig** [comment]
When I create a section current the POST call is failing? may be some integration issue?

**Comment by pushyamig** [comment]
1. we should validate no duplicate email address are there. 

Screen Shot 2021-12-17 at 4 40 12 PM

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:45]: That's a component I'm not touching, so maybe this can be a separate issue. We should probably add validation to that field too.

Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:45]: Agree though, FWIW

**Comment by ssciolla** [comment]
> 1. we should validate no duplicate email address are there.
> 
> 
> Screen Shot 2021-12-17 at 4 40 12 PM

Yep, I knew that was something I should do, I'll add it to the to-do list.

**Comment by ssciolla** [comment]
> When I create a section current the POST call is failing? may be some integration issue?

I couldn't reproduce that, seemed to work for me. What does it say in the console?

File: ccm_web/client/src/models/canvas.ts
Comment (pushyamig) [inline ccm_web/client/src/models/canvas.ts:167]: I like the way you designed this role assigning object. 
Screen Shot 2021-12-17 at 4 51 53 PM


Few violation when populating roles list based on canvas role
1. Teacher/Designer should able to add person with teacher role
2.  Assistant has no roles populated in dropdown

Comment (ssciolla) [reply ccm_web/client/src/models/canvas.ts:167]: Thanks, based on your feedback though, I may need to adjust the logic. Because teachers are also allowed to add teachers, the `rank` system doesn't quite work. I may just change it to having a few different role sets and assigning them manually to each of the roles that can add.

Comment (ssciolla) [reply ccm_web/client/src/models/canvas.ts:113]: Okay, this is what I arrived at, but I'll review it again to make sure it makes sense: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/d94e518dc364d3e0b646a0110a31c7466e1f841e

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:48]: I think I took care of this actually, it made sense to do it in the composite component you commented on.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (ssciolla) [inline ccm_web/client/src/components/CreateSelectSectionWidget.tsx:46]: Maybe we should revisit the helper text here. I'd replace "distinct" with "unique" or explain it differently.

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (ssciolla) [inline ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:233]: The `RoleRowsValidator` uses `isValidRole`, which is also a type guard, but that type information isn't available here. I think based on the context and logic, it's safe to type cast, but there may be a better way to do it.

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (ssciolla) [inline ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:318]: I'll revisit this when the API calls are in place.

File: ccm_web/client/src/components/RoleSelect.tsx
Comment (ssciolla) [inline ccm_web/client/src/components/RoleSelect.tsx:33]: How do we plan to document for user on which role/enrollment types they can use? I've been using "Only roles less privileged than your own are allowed", but that actually isn't accurate, since instructors I guess can add other instructors. I think we could say "See the docs for info on which roles you can use". I guess this is more important for the bulk process, since we will need to throw a validation error if they provide one they can't use. Here, we'll just show them what they can, although they may have questions about their options.

File: ccm_web/client/src/models/canvas.ts
Comment (ssciolla) [inline ccm_web/client/src/models/canvas.ts:68]: I don't 100 percent remember why I changed the names here. I don't think it's a major issue, but if someone feels strongly, I can change it back to `ClientRoleType` and `CanvasRoleType`. Maybe I was trying to distinguish between `RoleEnum` and `CanvasRoleType`. I wonder if there's a way to combine these enums, but that may be out of scope for this issue.

**Comment by ssciolla** [comment]
We're still waiting on UX input for what we discussed yesterday WRT the "Search" button, but I think this is mostly ready for review. I will finish the description and mark it as non-draft this afternoon.

**Comment by ssciolla** [comment]
Okay, I'm opening this up for review. It likely has flaws, and we're still discussing some usability aspects, but I think it's a step in the right direction. We can build on this as necessary, open new issues if needed. I may make small additional changes as I see fit, or incorporate usability feedback if the changes are not extensive.

**Comment by ssciolla** [comment]
I've assigned @lsloan and @zqian to review, but I'm not sure on their availability. @pushyamig can probably take point.

Apologies for letting this PR get so complex, with lots of new components. I guess my style is to refactor as I go, which results in PRs like this.

**Comment by ssciolla** [comment]
I ignore some Codacy issues, all of them are not present in VSCode. I believe they are due to the library typing issue discussed in #215.

**Comment by pushyamig** [comment]
Screen Shot 2022-01-06 at 3 50 08 PM

I see a validation error thrown when duplicate email address is added.

**Comment by ssciolla** [comment]
> I don't see a validation error thrown when duplicate email address is added.

Looking now...

**Comment by pushyamig** [comment]
>I don't see a validation error thrown when duplicate email address is added.
Looking now...

please ignore, I did not pull latest from you branch. as this was not a brand new fetch. Sorry for the confusion.

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/UserEnrollmentForm.tsx:383]: I feel there must be tinny line gap between feature title and feature option for adding single or multiple user
Screen Shot 2022-01-06 at 4 07 28 PM

Comment (ssciolla) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:383]: Do you want the spacing/gap to be larger?

Comment (pushyamig) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:383]: just a smidgen

Comment (ssciolla) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:383]: See https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/7b3a2aa7fc33923d08fff9acfd759551c44acc83

We should also consider adding similar spacing below the other `h1`s, but not sure if that should be done here.

I found I wasn't using this CSS class anymore, so I'm going to ensure I didn't leave any other stranded classes in other components.

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/UserEnrollmentForm.tsx:366]: Update as per @melindakraft `The new user was invited to choose a login method, added to Canvas, and enrolled in the selected section.`

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:312]: Update as per @melindakraft `Non-UM Users have been added to the designated sections. New users have been invited to choose a login method.`

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/UserEnrollmentForm.tsx:220]:
```suggestion
      Please provide a first and last name, and we will invite them to set up a login method.
```

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:160]:
```suggestion
      'If they do not have an account in Canvas, they will be sent an invitation to choose a login method and added to Canvas ' +
```

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (pushyamig) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:383]: >We should also consider adding similar spacing below the other h1s, but not sure if that should be done here.

I think the gap is not obvious with other features since none as of now as sub-selection. So Add UM will have that if you resolve #278

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:312]: Should it just be "Non-UM Users have been added to the selected section! New users have been sent an email invitation to choose a login method"? We're only supporting one section right now, I had worded it wrong before.

Comment (ssciolla) [reply ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:312]: Or really "Non-UM Users have been added to the selected section! New users have also been added to Canvas and sent an email invitation to choose a login method."

Comment (pushyamig) [reply ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx:312]: ok. I will leave it up to you further. my only thing here was friend account is just one aspect of creating Non-UM users. So you can rephrase as you like.

**Comment by pushyamig** [comment]
This seems to be functional from what I have tested
1. Email address to firstname/lastname validation and overall form validation are accurate
2. Role are displayed and validated appropriately across, instructor,TA.Designer, Assistant, Admin.  
3. Overall the look and feel is good.

Nice work @ssciolla. Tomorrow I will review from a coding perspective ( lot of files so i will do my best understanding it)

File: ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/MultipleUserEnrollmentWorkflow.tsx]: Done. See https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/be08a25d4f0268b266238d997c5c5324ccec07ac

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:367]: Done. See https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/be08a25d4f0268b266238d997c5c5324ccec07ac

File: ccm_web/client/src/pages/AddNonUMUsers.tsx
Comment (pushyamig) [inline ccm_web/client/src/pages/AddNonUMUsers.tsx:37]: I see that `single` and `csv` is used multiple place in this file. I feel it should be extracted as a constant and use it

File: ccm_web/client/src/pages/AddNonUMUsers.tsx
Comment (pushyamig) [inline ccm_web/client/src/pages/AddNonUMUsers.tsx:78]: For Future/Icebox: I feel that the L60-L78, can be extracted to as separate component and can be reused in features like where ever the sections list is needed.

File: ccm_web/client/src/pages/AddNonUMUsers.tsx
Comment (pushyamig) [inline ccm_web/client/src/pages/AddNonUMUsers.tsx:163]: The code in this file is very well organized, easy to read and separating each selection to a component is really neat.

File: ccm_web/client/src/components/UserEnrollmentForm.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/UserEnrollmentForm.tsx:71]: `loginId` is not being used in this block. should you remove it?

Comment (pushyamig) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:71]: or probably you will use it when you actually make a real API Call? in that case you can leave it

Comment (ssciolla) [reply ccm_web/client/src/components/UserEnrollmentForm.tsx:71]: Yeah, it will be used later.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (pushyamig) [inline ccm_web/client/src/components/CreateSelectSectionWidget.tsx:1]: Really nice that you separated this as component and can be reused for future features!

File: ccm_web/client/src/components/GradebookCanvasValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/components/GradebookCanvasValidators.ts:1]: Optional: I saw a couple of files as Validators, could go inside a sub-folder in component as `validator` for code organization purpose.

Comment (ssciolla) [reply ccm_web/client/src/components/GradebookCanvasValidators.ts:1]: That would be great, I'll consider doing that in the future, but I didn't want to change even more in this PR.

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/utils/enrollmentValidators.ts:37]: can you make the `email address` as  const here since the string is used in more that 2 places

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/utils/enrollmentValidators.ts:25]: can you add a comment why `i+2`

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [reply ccm_web/client/src/utils/enrollmentValidators.ts:37]: Optional: same goes for `first name, last name, loginID` but since only one file that is fine leave off

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/utils/enrollmentValidators.ts:72]: This block is repeated for most/ all validation, could this be generalized?

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/utils/enrollmentValidators.ts:1]: Some of the repetitive code in this file could be generalized see my comment https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/files#r780433576

File: ccm_web/client/src/models/models.ts
Comment (pushyamig) [inline ccm_web/client/src/models/models.ts:100]: Really Nice on extracting and reusing this across features

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts:72]: Yeah, I thought about this, let me give it a shot.

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts:1]: How's this? https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/7238bc99ebf59b82dfaa6fefb40aeb045d37368b

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts]: How's this? https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/7238bc99ebf59b82dfaa6fefb40aeb045d37368b

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (pushyamig) [reply ccm_web/client/src/utils/enrollmentValidators.ts:1]: Nice!

File: ccm_web/client/src/components/BulkSectionCreateValidators.ts
Comment (pushyamig) [inline ccm_web/client/src/components/BulkSectionCreateValidators.ts:3]: There is an linting error saying `validateString` not being used.

Comment (ssciolla) [reply ccm_web/client/src/components/BulkSectionCreateValidators.ts]: Fixed. https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/6a05e07b312bfa996844678ccbe0fcfca46c4441

**Review by pushyamig — APPROVED** [review]
Apart from the Non-UM user UI development, the code also refactored that touched multiple features. Tested overall features and things seems to be working fine. 

This is huge code base I did my diligence in understanding, reviewing and testing. Overall I feel good with the new feature development and Refactor.

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:46]: @pushyamig, thoughts on this?

File: ccm_web/client/src/components/RoleSelect.tsx
Comment (ssciolla) [reply ccm_web/client/src/components/RoleSelect.tsx]: @pushyamig, what's our plan here?

File: ccm_web/client/src/components/CreateSelectSectionWidget.tsx
Comment (pushyamig) [reply ccm_web/client/src/components/CreateSelectSectionWidget.tsx:46]: Keep it as it is. I don't have any opinion.

File: ccm_web/client/src/pages/AddNonUMUsers.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/AddNonUMUsers.tsx:78]: See issue #296.

File: ccm_web/client/src/components/RoleSelect.tsx
Comment (pushyamig) [reply ccm_web/client/src/components/RoleSelect.tsx]: I don't think the second line is needed and is communicating wrong info too.  I think for single add user case the dropdown is clear indicator what role he can add. 
 I see the CSV feature is validating if user is trying to add a role if he is not supposed to. We might provide this info in documentation

I won't hold this issue back just for deciding on this.

Comment (ssciolla) [reply ccm_web/client/src/components/RoleSelect.tsx]: I've already updated the text to be more accurate:
```
    <Typography id='role-select-label' className={classes.spacing}>
      Select a Canvas role that the user will have in the section.
      The available roles are determined by your own role(s) in the course.
    </Typography>
```

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts]: It's `i + 2` because `i` is an array index and JS uses zero-index arrays, so we add + 1. The column is not considered a data row, so we add + 1 again.

Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts]: Oh, you mean in the code, sure I can add a comment.

Comment (pushyamig) [reply ccm_web/client/src/utils/enrollmentValidators.ts]: yes

Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts]: Okay, as discussed via DM, I just extracted this calculation as `getRowNumber` so I didn't have to repeat this comment everywhere. https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/8bbec4409aff223262e1d16898be8a7e84ff41a4

This has minimal utility, but may help us keep things consistent or improve readability. And I don't have to repeat the same comments.

File: ccm_web/client/src/utils/enrollmentValidators.ts
Comment (ssciolla) [reply ccm_web/client/src/utils/enrollmentValidators.ts:22]: Can I postpone doing this, @pushyamig? It may make sense to put these constants in a different file, since I'm already using them in `validation`. Or maybe it makes sense to pair the `yup` schemas with the string, like this: 

```
interface FieldSchemaData {
  fieldName: string,
  schema: StringSchema
}
```

Anyways, this doesn't feel like it should be a blocker to merging this, right?

Comment (pushyamig) [reply ccm_web/client/src/utils/enrollmentValidators.ts:22]: Sure.

File: ccm_web/client/src/pages/AddNonUMUsers.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/AddNonUMUsers.tsx]: Okay, @pushyamig, I think I've kind of resolved this by using a string enum. What do you think? https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/270/commits/4131f2871bc83e46cd6698b6db47e5cf629f2598

Comment (pushyamig) [reply ccm_web/client/src/pages/AddNonUMUsers.tsx]: Nice

**Comment by ssciolla** [comment]
Okay, I'm going to merge now. Thank you, @pushyamig, for your patience and weeding through all of my code. I may have missed something, but we'll keep looking at this I'm sure when the backend is integrated.

**Pull request merged**
